File: billing-service/src/main/java/com/hotelpms/billing/service/KpiReportService.java
package com.hotelpms.billing.service;

import com.hotelpms.billing.dto.KpiReportDto;
import com.hotelpms.billing.dto.ReportGranularity;

import java.time.LocalDate;
import java.util.UUID;

/**
 * Service for computing the RevPAR/ADR/Occupancy trend report (epic C4).
 */
@FunctionalInterface
public interface KpiReportService {

    /**
     * Computes the KPI trend for a date range, scoped to a hotel.
     *
     * @param hotelId     the hotel UUID (tenant isolation)
     * @param startDate   start of the period (inclusive)
     * @param endDate     end of the period (inclusive)
     * @param granularity the time-bucket size
     * @return the KPI report
     */
    KpiReportDto getKpiReport(UUID hotelId, LocalDate startDate, LocalDate endDate, ReportGranularity granularity);
}

File: frontend/src/services/kpiReportService.ts
import api from './api';
import type { KpiReportDto, ReportGranularity } from '../types/ownerReport.types';

const KPI_REPORT_PATH = '/api/v1/reports/kpi';

export const kpiReportService = {
  getKpiReport: async (
    startDate: string,
    endDate: string,
    granularity: ReportGranularity
  ): Promise<KpiReportDto> => {
    const response = await api.get<KpiReportDto>(KPI_REPORT_PATH, {
      params: { startDate, endDate, granularity },
    });
    return response.data;
  },
};
